README


This archive contains the distribution Etcd,
version 0.003:

  Client library for etcd

This software is copyright (c) 2014 by Robert Norris.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.


This README file was generated by Dist::Zilla::Plugin::Readme v5.019.



Changes
0.003     2014-10-21 08:17:12+11:00 Australia/Melbourne
 - handle JSON::XS booleans when creating response objects
 - explicitly require HTTP::Tiny >= 0.014

0.002     2014-10-14 10:00:41+11:00 Australia/Melbourne
 - Minor doc fixes
 - Handle all redirects to leader nodes
 - More tests

0.001     2014-06-17 22:54:50+10:00 Australia/Melbourne
 - First release


lib/Etcd.pm
package Etcd;
$Etcd::VERSION = '0.003';
# ABSTRACT: Client library for etcd

use namespace::sweep;

use HTTP::Tiny 0.014;
use URI::Escape qw(uri_escape);
use Carp qw(croak);

use Moo;
use Type::Utils qw(class_type);
use Types::Standard qw(Str Int Bool HashRef);

has host => ( is => 'ro', isa => Str, default => '127.0.0.1' );
has port => ( is => 'ro', isa => Int, default => 4001 );

has ssl => ( is => 'ro', isa => Bool, default => 0 );

has http => ( is => 'lazy', isa => class_type('HTTP::Tiny') );
sub _build_http { HTTP::Tiny->new };

has version_prefix => ( is => 'ro', isa => Str, default => '/v2' );

has _url_base => ( is => 'lazy' );
sub _build__url_base {
    my ($self) = @_;
    ($self->ssl ? 'https' : 'http') .'://'.$self->host.':'.$self->port;
}

sub _prep_url {
    my ($self, $path, %args) = @_;
    my $trailing = $path =~ m{/$};
    my $url = $self->_url_base.join('/', map { uri_escape($_) } split('/', $path));
    $url .= '/' if $trailing;
    $url .= '?'.$self->http->www_form_urlencode(\%args) if %args;
    $url;
}

sub api_exec {
    my ($self, $path, $method, %args) = @_;
    my $res = $self->http->request($method, $self->_prep_url($path, %args));
    $res = $self->http->request($method, $res->{headers}->{location})
        if $res && $res->{status} eq 307;
    return $res if $res->{success};
    croak "$res->{status} $res->{reason}: $res->{content}" if $res->{status} >= 500;
    require Etcd::Error;
    die Etcd::Error->new_from_http($res);
}

sub server_version {
    my ($self) = @_;
    my $res = $self->api_exec('/version', 'GET');
    $res->{content};
}

with 'Etcd::Keys';
with 'Etcd::Stats';

1;

__END__

=pod

=encoding UTF-8

=head1 NAME

Etcd - Client library for etcd

=head1 VERSION

version 0.003

=head1 SYNOPSIS

    use Etcd;
    my $etcd = Etcd->new;
    
    say "server version: ".$etcd->server_version;
    
    # Key space API
    # See Etcd::Keys for more info
    $etcd->set("/message", "hello world");
    say "message is: ".$etcd->get("/message")->node->value;
    
    # Stats API
    # See Etcd::Stats for more info
    say "leader is: ".$etcd->stats("leader")->{leader};

=head1 DESCRIPTION

This is a client library for accessing and manipulating data in an etcd
cluster. It targets the etcd v2 API.

This module is quite low-level. You're expected to have a good understanding of
etcd and its API to understand the methods this module provides. See L</SEE
ALSO> for further reading.

=head1 METHODS

=head2 new

    my $etcd = Etcd->new( %args );

This constructor returns a new Etcd object. Valid arguments include:

=over 4

=item *

C<host>

Hostname or IP address of an etcd server (default: C<127.0.0.1>)

=item *

C<port>

Port where the etcd server is listening (default: C<4001>)

=item *

C<ssl>

Use SSL/TLS (ie HTTPS) when talking to the etcd server (default: off)

=item *

C<http>

A C<HTTP::Tiny> object to use to access the server. If not specified, one will
be created

=back

=head2 server_version

    my $version = $etcd->server_version;

Queries and returns the server version as a string.

=head1 ENDPOINTS

Individual API endpoints are implemented in separate modules. See the documentation for the following modules for more information:

=over 4

=item *

L<Etcd::Keys> - Key space API

=item *

L<Etcd::Stats> - Stats API

=back

=head1 SEE ALSO

=over 4

=item *

L<HTTP::Tiny> - for further HTTP client configuration, especially SSL configuration

=item *

L<https://github.com/coreos/etcd> - etcd source and documentation

=item *

L<https://coreos.com/docs/distributed-configuration/etcd-api/> - etcd API documentation

=back

=head1 SUPPORT

=head2 Bugs / Feature Requests

Please report bugs or feature requests through the issue tracker at
L<https://github.com/robn/p5-etcd/issues>. You will be notified automatically
of any progress on your issue.

=head2 Source Code

This is open source software.  The code repository is available for
public review and contribution under the terms of the license.

L<https://github.com/robn/p5-etcd>

  git clone https://github.com/robn/p5-etcd.git

=head1 AUTHORS

=over 4

=item *

Robert Norris <rob@eatenbyagrue.org>

=item *

Matt Harrington

=back

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2014 by Robert Norris.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut


lib/Etcd/Keys.pm
package Etcd::Keys;
$Etcd::Keys::VERSION = '0.003';
use namespace::sweep;

use Etcd::Response;
use Try::Tiny;
use Scalar::Util qw(blessed);
use Carp qw(croak);

use Moo::Role;
use Types::Standard qw(Str);

requires qw(version_prefix api_exec);

has _keys_endpoint => ( is => 'lazy', isa => Str );
sub _build__keys_endpoint {
    shift->version_prefix . '/keys';
}

sub set {
    my ($self, $key, $value, %args) = @_;
    croak 'usage: $etcd->set($key, $value, [%args])' if grep { !defined } ($key, $value);
    Etcd::Response->new_from_http($self->api_exec($self->_keys_endpoint.$key, 'PUT', %args, value => $value));
}

sub get {
    my ($self, $key, %args) = @_;
    croak 'usage: $etcd->get($key, [%args])' if !defined $key;
    Etcd::Response->new_from_http($self->api_exec($self->_keys_endpoint.$key, 'GET', %args));
}

sub delete {
    my ($self, $key, %args) = @_;
    croak 'usage: $etcd->delete($key, [%args])' if !defined $key;
    Etcd::Response->new_from_http($self->api_exec($self->_keys_endpoint.$key, 'DELETE', %args));
}

sub compare_and_swap {
    my ($self, $key, $value, $prev_value, %args) = @_;
    croak 'usage: $etcd->compare_and_swap($key, $value, $prev_value, [%args])' if grep { !defined } ($key, $value, $prev_value);
    $self->set($key, $value, %args, prevValue => $prev_value);
}

sub compare_and_delete {
    my ($self, $key, $prev_value, %args) = @_;
    croak 'usage: $etcd->compare_and_delete($key, $prev_value, [%args])' if grep { !defined } ($key, $prev_value);
    $self->delete($key, %args, prevValue => $prev_value);
}

sub create {
    my ($self, $key, $value, %args) = @_;
    croak 'usage: $etcd->create($key, $value, [%args])' if grep { !defined } ($key, $value);
    $self->set($key, $value, %args, prevExist => 'false');
}

sub update {
    my ($self, $key, $value, %args) = @_;
    croak 'usage: $etcd->update($key, $value, [%args])' if grep { !defined } ($key, $value);
    $self->set($key, $value, %args, prevExist => 'true');
}

sub exists {
    my ($self, $key, %args) = @_;
    croak 'usage: $etcd->exists($key, [%args])' if !defined $key;
    try {
        $self->get($key, %args);
        1;
    }
    catch {
        die $_ unless defined blessed $_ && $_->isa('Etcd::Error');
        die $_ unless $_->error_code == 100;
        "";
    }
}

sub create_dir {
    my ($self, $key, %args) = @_;
    croak 'usage: $etcd->create_dir($key, [%args])' if !defined $key;
    Etcd::Response->new_from_http($self->api_exec($self->_keys_endpoint.$key, 'PUT', %args, dir => 'true'));
}

sub delete_dir {
    my ($self, $key, %args) = @_;
    croak 'usage: $etcd->delete_dir($key, [%args])' if !defined $key;
    $self->delete($key, %args, dir => 'true');
}

sub create_in_order {
    my ($self, $key, $value, %args) = @_;
    croak 'usage: $etcd->create_in_order($key, $value, [%args])' if grep { !defined } ($key, $value);
    Etcd::Response->new_from_http($self->api_exec($self->_keys_endpoint.$key, 'POST', %args, value => $value));
}

sub watch {
    my ($self, $key, %args) = @_;
    croak 'usage: $etcd->watch($key, [%args])' if !defined $key;
    $self->get($key, %args, wait => 'true');
}

1;

__END__

=pod

=encoding UTF-8

=head1 NAME

Etcd::Keys - etcd key space API

=head1 VERSION

version 0.003

=head1 SYNOPSIS

    use Etcd;
    my $etcd = Etcd->new;
    
    # set value for key
    $etcd->set("/message", "hello world");
    
    # get key
    my $response = $etcd->get("/message");
    
    # delete key
    $etcd->delete("/message");
    
    # atomic compare-and-swap value for key
    $etcd->compare_and_swap("/message", "new", "old");
    
    # atomic compare-and-delete key
    $etcd->compare_and_delete("/message", "old");
    
    # create key. like set, but fails if the key exists
    $etcd->create("/message", "value");
    
    # update key. like set, but fails if the key doesn't exist
    $etcd->update("/message", "value");
    
    # check if key exists
    my $exists = $etcd->exists("/message");
    
    # create dir, a "valueless" key to hold subkeys
    $etcd->create_dir("/dir");
    
    # delete key and everything under it
    $etcd->delete_dir("/dir");
    
    # atomically create in-order key
    $etcd->create_in_order("/dir", "value");
    
    # block until key changes
    $etcd->watch("/message");

=head1 DESCRIPTION

This module provides access to etcd's key space API. The methods here map
almost exactly to operations described in the etcd API documentation. See
L<Etcd/SEE ALSO> for further reading.

All methods except C<exists> returns a L<Etcd::Response> object on success and
C<die> on error. On error, C<$@> will contain either a reference to a
L<Etcd::Error> object (for API-level errors) or a regular string (for network,
transport or other errors).

All methods can take any number of additional arguments in C<key =E<gt> value>
form. These are added to the query parameters in the URL that gets submitted to
etcd. This is how you would pass options for C<ttl> or C<recursive>, for
example.

Any arguments of this kind that clash with the internal operation of a method
will silently be ignored; for example, passing a C<value> key to C<set> will be
ignored because that's how the value is passed internally.

=head1 METHODS

=over 4

=item *

C<set>

    $etcd->set("/message", "hello world");

Set a value for a key. The key will be created if it doesn't exist.

This invokes the C<PUT> method for the given key.

=item *

C<get>

    my $node = $etcd->get("/message");

Get a key.

This invokes the C<GET> method for the given key.

=item *

C<delete>

    $etcd->delete("/message");

Delete a key.

This invokes the C<DELETE> method for the given key.

=item *

C<compare_and_swap>

    $etcd->compare_and_swap("/message", "new", "old");

Atomic compare-and-swap the value of a key.

This invokes the C<PUT> method for the given key with the C<prevValue> query
parameter.

=item *

C<compare_and_delete>

    $etcd->compare_and_delete("/message", "old");

Atomic compare-and-delete the value of a key.

This invokes the C<DELETE> method for the given key with the C<prevValue> query
parameter.

=item *

C<create>

    $etcd->create("/message", "value");

Create a key. Like set, but fails if the key exists.

This invokes the C<PUT> method for the given key with the C<prevExist> query
parameter set to C<false>.

=item *

C<update>

    $etcd->update("/message", "value");

Update the value of a key. Like set, but fails if the key doesn't exist.

This invokes the C<PUT> method for the given key with the C<prevExist> query
parameter set to C<true>.

=item *

C<exists>

    my $exists = $etcd->exists("/message");

Check if key exists. Unlike the other methods, it does not return a reference
to a L<Etcd::Response> object but insteads returns a true or false value. It
may still throw an error.

This invokes the C<GET> method for the given key.

=item *

C<create_dir>

    $etcd->create_dir("/dir");

Creates a directory, a "valueless" key to hold sub-keys.

This invokes the C<PUT> method for the given key with the C<dir> query
parameter set to C<true>.

=item *

C<delete_dir>

    $etcd->delete_dir("/dir");

Deletes a key and all its sub-keys.

This invokes the C<DELETE> method for the given key with the C<dir> query
parameter set to C<true>.

=item *

C<create_dir>

    $etcd->create_in_order("/dir", "value");

Atomically creates an in-order key.

This invokes the C<POST> method for the given key.

=item *

C<watch>

    $etcd->watch("/message");

Block until the given key changes, then return the change.

This invokes the C<GET> method for the given key with the C<wait> query
parameter set to C<true>.

=back

=head1 KNOWN ISSUES

=over 4

=item *

There is no convenient way to specify the C<prevIndex> test to
C<compare_and_swap> or C<compare_and_delete>. These can be implemented directly
with C<set>.

=item *

C<watch> has no asynchronous mode.

=back

See L<Etcd/SUPPORT> for information on how to report bugs or feature requests.

=head1 AUTHORS

=over 4

=item *

Robert Norris <rob@eatenbyagrue.org>

=back

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2014 by Robert Norris.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut


lib/Etcd/Node.pm
package Etcd::Node;
$Etcd::Node::VERSION = '0.003';
use namespace::sweep;

use Moo;
use Type::Tiny;
use Types::Standard qw(Int Str Bool ArrayRef);
use Type::Utils qw(class_type);

# http://www.pelagodesign.com/blog/2009/05/20/iso-8601-date-validation-that-doesnt-suck/
my $ISO8601 = do {
    my $iso8601_re = qr{^([\+-]?\d{4}(?!\d{2}\b))((-?)((0[1-9]|1[0-2])(\3([12]\d|0[1-9]|3[01]))?|W([0-4]\d|5[0-2])(-?[1-7])?|(00[1-9]|0[1-9]\d|[12]\d{2}|3([0-5]\d|6[1-6])))([T\s]((([01]\d|2[0-3])((:?)[0-5]\d)?|24\:?00)([\.,]\d+(?!:))?)?(\17[0-5]\d([\.,]\d+)?)?([zZ]|([\+-])([01]\d|2[0-3]):?([0-5]\d)?)?)?)?$};
    Type::Tiny->new(
        name       => "ISO8601",
        constraint => sub { m/$iso8601_re/ },
        inlined    => sub { "$_[1] =~ m/$iso8601_re/" },
    );
};

has key            => ( is => 'ro', isa => Str, required => 1 );
has value          => ( is => 'ro', isa => Str );
has created_index  => ( is => 'ro', isa => Int, init_arg => 'createdIndex'  );
has modified_index => ( is => 'ro', isa => Int, init_arg => 'modifiedIndex' );
has ttl            => ( is => 'ro', isa => Int );
has expiration     => ( is => 'ro', isa => $ISO8601 );
has dir            => ( is => 'ro', isa => Bool, coerce => sub { !! $_[0] } );

has nodes => (
    is => 'ro',
    isa => ArrayRef[class_type('Etcd::Node')],
    coerce => sub { ref $_[0] eq "ARRAY" ? [ map { Etcd::Node->new(%$_) } @{$_[0]} ] : $_[0] }
);

1;

__END__

=pod

=encoding UTF-8

=head1 NAME

Etcd::Node - key space node representation

=head1 VERSION

version 0.003

=head1 SYNOPSIS

    use Etcd;
    my $etcd = Etcd->new;
    
    my $node = $etcd->get("/message")->node;

=head1 DESCRIPTION

L<Etcd::Node> objects encapsulate the details of nodes in the key space API.

The provided methods are simple accessors. This class provides no
functionality.

The API docs have more information about the meaning of each item. See
L<Etcd/SEE ALSO> for further reading.

=head1 METHODS

=over 4

=item *

C<key>

=item *

C<value>

Value of this key. Can be C<undef> (usually in directory nodes).

=item *

C<created_index>

=item *

C<modified_index>

=item *

C<ttl>

=item *

C<expiration>

ISO-8601 string indicating the moment that this node will be deleted.

=item *

C<dir>

True if the node is a directory. C<value> will undef in this case.

=item *

C<nodes>

Arrayref of zero or more sub-nodes. Only returned from appropriate
L<Etcd::Keys> calls with the C<recursive> parameter set to C<true>.

=back

=head1 AUTHORS

=over 4

=item *

Robert Norris <rob@eatenbyagrue.org>

=back

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2014 by Robert Norris.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut


lib/Etcd/Error.pm
package Etcd::Error;
$Etcd::Error::VERSION = '0.003';
use namespace::sweep;

use JSON qw(decode_json);
use Carp qw(longmess);

use Moo;
use Types::Standard qw(Str Int);

has error_code => ( is => 'ro', isa => Str, required => 1, init_arg => 'errorCode' );
has message    => ( is => 'ro', isa => Str, required => 1 );
has cause      => ( is => 'ro', isa => Str, required => 1 );
has index      => ( is => 'ro', isa => Int, required => 1 );
has trace      => ( is => 'ro', isa => Str, required => 1 );

sub new_from_http {
    my ($class, $res) = @_;
    my $data = decode_json($res->{content});
    $class->new(%$data, trace => longmess());
}

use overload
    q{""} => sub {
        my ($self) = @_;
        $self->cause.": ".$self->message." [".$self->error_code."]".$self->trace;
    };

1;

__END__

=pod

=encoding UTF-8

=head1 NAME

Etcd::Error - API error representation

=head1 VERSION

version 0.003

=head1 SYNOPSIS

    use Etcd;
    my $etcd = Etcd->new;
    
    use Try::Tiny;
    try {
        $etcd->get("/message");
    }
    catch {
        print $_;
    };

=head1 DESCRIPTION

L<Etcd::Error> objects encapsulate the details of API errors. They are thrown
by API calls when something goes wrong.

The provided methods are simple accessors. A stringification overload is
provided to produce a meaningful error with backtrace.

The API docs have more information about the meaning of each item. See
L<Etcd/SEE ALSO> for further reading.

=head1 METHODS

=over 4

=item *

C<error_code>

=item *

C<message>

=item *

C<cause>

=item *

C<index>

=item *

C<trace>

Stacktrace from the point that the error was generated.

=back

=head1 AUTHORS

=over 4

=item *

Robert Norris <rob@eatenbyagrue.org>

=back

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2014 by Robert Norris.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut


lib/Etcd/Stats.pm
package Etcd::Stats;
$Etcd::Stats::VERSION = '0.003';
use namespace::sweep;

use JSON qw(decode_json);
use Carp qw(croak);

use Moo::Role;
use Types::Standard qw(Str);

requires qw(version_prefix api_exec);

has _stats_endpoint => ( is => 'lazy', isa => Str );
sub _build__stats_endpoint {
    shift->version_prefix . '/stats';
}

# XXX should return real objects
sub stats {
    my ($self, $type, %args) = @_;
    croak 'usage: $etcd->type("leader|store|self", [%args])' if !defined $type || $type !~ m/^(?:leader|store|self)$/;
    decode_json($self->api_exec($self->_stats_endpoint."/$type", 'GET', %args)->{content});
}

1;

__END__

=pod

=encoding UTF-8

=head1 NAME

Etcd::Stats - etcd stats API

=head1 VERSION

version 0.003

=head1 SYNOPSIS

    use Etcd;
    my $etcd = Etcd->new;
    
    my $leader_stats = $etcd->stats("leader");
    
    my $store_stats = $etcd->stats("store");
    
    my $self_stats = $etcd->stats("self");

=head1 DESCRIPTION

This module provides access to etcd's stats API.

=head1 METHODS

=over 4

=item *

C<stats>

    my $leader_stats = $etcd->stats("leader");

    my $store_stats = $etcd->stats("store");

    my $self_stats = $etcd->stats("self");

Returns a hashref of values for the requested statistics type. The contents of
the hash is described in the API documentation. See L<Etcd/SEE ALSO> for
further reading.

On error, C<$@> will contain either a reference to a L<Etcd::Error> object (for
API-level errors) or a regular string (for network, transport or other errors).

=back

=head1 AUTHORS

=over 4

=item *

Robert Norris <rob@eatenbyagrue.org>

=back

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2014 by Robert Norris.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut


lib/Etcd/Response.pm
package Etcd::Response;
$Etcd::Response::VERSION = '0.003';
use namespace::sweep;

use Etcd::Node;

use JSON qw(decode_json);

use Moo;
use Type::Utils qw(class_type);
use Types::Standard qw(Str Int);

has action     => ( is => 'ro', isa => Str, required => 1 );
has etcd_index => ( is => 'ro', isa => Int, required => 1 );
has raft_index => ( is => 'ro', isa => Int, required => 1 );
has raft_term  => ( is => 'ro', isa => Int, required => 1 );

my $node_coercion = sub { ref $_[0] eq 'HASH' ? Etcd::Node->new(%{$_[0]}) : $_[0] };
has node       => ( is => 'ro', isa => class_type('Etcd::Node'), coerce => $node_coercion, required => 1 );
has prev_node  => ( is => 'ro', isa => class_type('Etcd::Node'), coerce => $node_coercion, init_arg => 'prevNode' );

sub new_from_http {
    my ($class, $res) = @_;
    my $data = decode_json($res->{content});
    my %headers;
    @headers{qw(etcd_index raft_index raft_term)} = @{$res->{headers}}{qw(x-etcd-index x-raft-index x-raft-term)};
    $class->new(%$data, %headers);
}

1;

__END__

=pod

=encoding UTF-8

=head1 NAME

Etcd::Response - response from key space API

=head1 VERSION

version 0.003

=head1 SYNOPSIS

    use Etcd;
    my $etcd = Etcd->new;
    
    my $response = $etcd->get("/message");
    say $response->node->value;

=head1 DESCRIPTION

L<Etcd::Response> objects encapsulate the details of responses from the key
space API.

The provided methods are simple accessors. This class provides no
functionality.

The API docs have more information about the meaning of each item. See
L<Etcd/SEE ALSO> for further reading.

=head1 METHODS

=over 4

=item *

C<action>

=item *

C<etcd_index>

=item *

C<raft_index>

=item *

C<raft_term>

=item *

C<node>

A L<Etcd::Node> object with the node data from this response.

=item *

C<prev_node>

A L<Etcd::Node> object with the previous state of the node data from this response. May be C<undef>.

=back

=head1 AUTHORS

=over 4

=item *

Robert Norris <rob@eatenbyagrue.org>

=back

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2014 by Robert Norris.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut


Makefile.PL

# This file was automatically generated by Dist::Zilla::Plugin::MakeMaker v5.019.
use strict;
use warnings;



use ExtUtils::MakeMaker 6.30;



my %WriteMakefileArgs = (
  "ABSTRACT" => "Client library for etcd",
  "AUTHOR" => "Robert Norris <rob\@eatenbyagrue.org>",
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => "6.30"
  },
  "DISTNAME" => "Etcd",
  "EXE_FILES" => [],
  "LICENSE" => "perl",
  "NAME" => "Etcd",
  "PREREQ_PM" => {
    "Carp" => 0,
    "HTTP::Tiny" => "0.014",
    "JSON" => 0,
    "Moo" => 0,
    "Scalar::Util" => 0,
    "Try::Tiny" => 0,
    "Type::Tiny" => 0,
    "Type::Utils" => 0,
    "Types::Standard" => 0,
    "URI::Escape" => 0,
    "namespace::sweep" => 0
  },
  "TEST_REQUIRES" => {
    "Test::More" => 0
  },
  "VERSION" => "0.003",
  "test" => {
    "TESTS" => "t/*.t"
  }
);


my %FallbackPrereqs = (
  "Carp" => 0,
  "HTTP::Tiny" => "0.014",
  "JSON" => 0,
  "Moo" => 0,
  "Scalar::Util" => 0,
  "Test::More" => 0,
  "Try::Tiny" => 0,
  "Type::Tiny" => 0,
  "Type::Utils" => 0,
  "Types::Standard" => 0,
  "URI::Escape" => 0,
  "namespace::sweep" => 0
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  delete $WriteMakefileArgs{TEST_REQUIRES};
  delete $WriteMakefileArgs{BUILD_REQUIRES};
  $WriteMakefileArgs{PREREQ_PM} = \%FallbackPrereqs;
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);



